lib/graphql/schema/validator/required_validator.rb

```diff
@@ -51,27 +51,34 @@ def initialize(one_of: nil, argument: nil, message: nil, **default_options)
         end
 
         def validate(_object, context, value)
-          matched_conditions = 0
+          fully_matched_conditions = 0
+          partially_matched_conditions = 0
 
           if !value.nil?
             @one_of.each do |one_of_condition|
               case one_of_condition
               when Symbol
                 if value.key?(one_of_condition)
-                  matched_conditions += 1
+                  fully_matched_conditions += 1
                 end
               when Array
-                if one_of_condition.all? { |k| value.key?(k) }
-                  matched_conditions += 1
-                  break
+                full_match = one_of_condition.all? { |k| value.key?(k) }
+                partial_match = !full_match && one_of_condition.any? { |k| value.key?(k) }
+
+                if full_match
+                  fully_matched_conditions += 1
+                end
+
+                if partial_match
+                  partially_matched_conditions += 1
                 end
               else
                 raise ArgumentError, "Unknown one_of condition: #{one_of_condition.inspect}"
               end
             end
           end
 
-          if matched_conditions == 1
+          if fully_matched_conditions == 1 && partially_matched_conditions == 0
```

Comment:
A partial match alongside a full match violates a stricter definition of "exactly one" and "mutually exclusive".